Sources/PackageModel/PackageIdentity.swift

```diff
@@ -0,0 +1,230 @@
+            .compactMap { $0.removingPercentEncoding ?? String($0) }
+
+        var lastPathComponent = components.popLast() ?? ""
+        lastPathComponent.removeSuffixIfPresent(".git")
+        components.append(lastPathComponent)
+
+        self.description = components.joined(separator: "/")
+        self.computedName = String(lastPathComponent)
+    }
+}
+
+extension PackageIdentity: Equatable, Comparable {
+    private func compare(to another: PackageIdentity) -> ComparisonResult {
+        return self.description.compare(another.description, options: [.caseInsensitive, .diacriticInsensitive])
+    }
+
+    public static func == (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.compare(to: rhs) == .orderedSame
+    }
+
+    public static func < (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.compare(to: rhs) == .orderedAscending
+    }
+
+    public static func > (lhs: PackageIdentity, rhs: PackageIdentity) -> Bool {
+        return lhs.compare(to: rhs) == .orderedDescending
+    }
+}
+
+extension PackageIdentity: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(self.description)
+    }
+}
+
+extension PackageIdentity: Codable {
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let description = try container.decode(String.self)
+        self.init(description)
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(self.description)
+    }
+}
+
+extension PackageIdentity: JSONMappable, JSONSerializable {
+    public init(json: JSON) throws {
+        guard case .string(let string) = json else {
+            throw JSON.MapError.typeMismatch(key: "", expected: String.self, json: json)
+        }
+
+        self.init(string)
+    }
+
+    public func toJSON() -> JSON {
+        return .string(self.description)
+    }
+}
+
+extension PackageIdentity: ExpressibleByStringLiteral {
+    public init(stringLiteral value: String) {
+        self.init(value)
+    }
+}
+
+// MARK: -
+
+private enum Scheme: String, CustomStringConvertible, CaseIterable {
+    case file
+    case ftp
+    case ftps
+    case git
+    case http
+    case https
+    case ssh
+
+    public var description: String {
+        return self.rawValue
+    }
+}
+
+#if os(Windows)
+fileprivate let isSeparator: (Character) -> Bool = { $0 == "/" || $0 == "\\" }
+#else
+fileprivate let isSeparator: (Character) -> Bool = { $0 == "/" }
+#endif
+
+private extension Character {
+    var isDigit: Bool {
+        isHexDigit && !isLetter
+    }
+}
+
+private extension String {
+    @discardableResult
+    mutating func removePrefixIfPresent<T: StringProtocol>(_ prefix: T) -> Bool {
+        guard hasPrefix(prefix) else { return false }
+        removeFirst(prefix.count)
+        return true
+    }
```

Comment:
consider moving to Basics module, seems generally useful

Reply:
Yes, this is one that I've wanted many times.  Same for suffix.